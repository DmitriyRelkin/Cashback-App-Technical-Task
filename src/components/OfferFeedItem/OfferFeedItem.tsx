import React from 'react';
import {View, Text, StyleSheet, TouchableOpacity, Image} from 'react-native';
import dayjs from "dayjs";
import {useNavigation} from "@react-navigation/native";
import {useSelector} from "react-redux";

export interface IOfferFeedItem {
  retailerLogo: string;
  title: string;
  expirationDate: string;
  description: string;
  cashbackAmount: number;
}

const OfferFeedItem: React.FC = ({ item }: {item: IOfferFeedItem}) => {
  const navigation = useNavigation();
  const { retailerLogo, title, expirationDate, description, cashbackAmount, id } = item;
  const { data } = useSelector((state: any) => state.cashback);

  const isClaimed = data.indexOf(id) !== -1;
  return (
    <TouchableOpacity
      style={[styles.card, {backgroundColor: isClaimed ? '#BDFFB1' : 'white'}]}
      onPress={() => navigation.navigate('OfferDetails', { offer: item })}>
      <View style={styles.header}>
        <Image source={{uri: retailerLogo}} style={styles.retailerImage} />
        <Text style={styles.title}>{title}</Text>
        <Text style={styles.expiration}>Expired on: {dayjs(expirationDate).format('MMM D, YYYY')}</Text>
      </View>
      <Text style={styles.description}>{description}</Text>
      {!isClaimed ? (
        <Text style={styles.cashback}>Cashback: {cashbackAmount}%</Text>
      ) : (
        <Text style={styles.cashback}>Cashback Applied!</Text>
      )}

    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  card: {
    padding: 16,
    marginVertical: 8,
    marginHorizontal: 16,
    borderRadius: 8,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 8,
  },
  header: {
    alignItems: 'center',
  },
  retailerImage: {
    width: 40,
    height: 40,
    borderRadius: 20,
    marginRight: 8,
  },
  title: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  expiration: {
    fontSize: 14,
    fontStyle: 'italic',
    color: '#810000',
    fontWeight: "700"
  },
  description: {
    fontSize: 16,
    marginVertical: 8,
  },
  cashback: {
    fontSize: 16,
    color: '#099d35',
    fontWeight: "700"
  },
});

export default OfferFeedItem;
