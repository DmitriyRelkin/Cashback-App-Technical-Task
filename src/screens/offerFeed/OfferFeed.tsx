import React from 'react';
import { Text, FlatList, StyleSheet, ActivityIndicator } from 'react-native';
import {useSelector} from "react-redux";
import OfferFeedItem from "../../components/OfferFeedItem/OfferFeedItem";

const OfferFeed = () => {
  const { data, loading, error } = useSelector((state: any) => state.offers);

  const renderOffer = ({ item }) => (
    <OfferFeedItem item={item} />
  )

  if (loading) {
    return <ActivityIndicator style={styles.loader} size="large" />;
  }

  if (error) {
    return <Text style={styles.error}>Failed to load offers</Text>;
  }

  return (
    <FlatList
      data={data}
      renderItem={renderOffer}
      keyExtractor={(item) => item.id}
      contentContainerStyle={styles.list}
    />
  );
};

const styles = StyleSheet.create({
  list: {
    padding: 16,
  },
  loader: {
    flex: 1,
    justifyContent: 'center',
  },
  error: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    color: 'red',
  },
  card: {
    backgroundColor: 'white',
    padding: 16,
    marginVertical: 8,
    marginHorizontal: 16,
    borderRadius: 8,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 8,
  },
  header: {
    alignItems: 'center',
  },
  retailerImage: {
    width: 40,
    height: 40,
    borderRadius: 20,
    marginRight: 8,
  },
  title: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  expiration: {
    fontSize: 14,
    fontStyle: 'italic',
    color: '#810000',
    fontWeight: "700"
  },
  description: {
    fontSize: 16,
    marginVertical: 8,
  },
  cashback: {
    fontSize: 16,
    color: '#099d35',
    fontWeight: "700"
  },
});

export default OfferFeed;
