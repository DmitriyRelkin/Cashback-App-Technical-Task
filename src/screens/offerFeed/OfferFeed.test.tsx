import React from 'react';
import renderer from 'react-test-renderer';
import { Text, ActivityIndicator, FlatList } from 'react-native';
import { useSelector } from 'react-redux';
import OfferFeed from "./OfferFeed";
import OfferFeedItem from "../../components/OfferFeedItem/OfferFeedItem";

// Mock necessary hooks and components
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

jest.mock('../../components/OfferFeedItem/OfferFeedItem', () => {
  return 'OfferFeedItem';
});

describe('OfferFeed', () => {
  const mockUseSelector = useSelector as jest.Mock;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders the component correctly', () => {
    mockUseSelector.mockReturnValue({ data: [], loading: false, error: null });
    const tree = renderer.create(<OfferFeed />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays a loader when loading', () => {
    mockUseSelector.mockReturnValue({ data: [], loading: true, error: null });
    const tree = renderer.create(<OfferFeed />);
    const loader = tree.root.findByType(ActivityIndicator);
    expect(loader).toBeTruthy();
  });

  it('renders a list of offers when data is available', () => {
    const mockData = [
      { id: '1', retailerLogo: 'https://example.com/logo1.png', title: 'Offer 1', expirationDate: '2024-12-31', description: 'Description 1', cashbackAmount: 5 },
      { id: '2', retailerLogo: 'https://example.com/logo2.png', title: 'Offer 2', expirationDate: '2024-12-31', description: 'Description 2', cashbackAmount: 10 },
    ];
    mockUseSelector.mockReturnValue({ data: mockData, loading: false, error: null });
    const tree = renderer.create(<OfferFeed />);
    const flatList = tree.root.findByType(FlatList);
    expect(flatList.props.data).toEqual(mockData);
    expect(flatList.props.renderItem({ item: mockData[0] })).toEqual(<OfferFeedItem item={mockData[0]} />);
  });
});
