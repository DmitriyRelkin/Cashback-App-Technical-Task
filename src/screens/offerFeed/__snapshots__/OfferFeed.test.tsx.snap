// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferFeed renders the component correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "padding": 16,
    }
  }
  data={[]}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View />
</RCTScrollView>
`;
